Source/DafnyCore/AST/SourceOrigin.cs

```diff
@@ -46,11 +64,66 @@ public int Length() {
 
   // TODO rename to Generated, and Token.NoToken to Token.Generated, and remove AutoGeneratedToken.
   public static IOrigin NoToken => Token.NoToken;
-  private readonly Token endToken;
+  private readonly Token? endToken;
+
+  public Token Center {
+    get;
+  }
+  public string TrailingTrivia {
+    get => Center.TrailingTrivia;
+    set => throw new InvalidOperationException();
+  }
+
+  public string LeadingTrivia {
+    get => Center.LeadingTrivia;
+    set => throw new InvalidOperationException();
+  }
+
+  public Token Next {
+    get => Center.Next;
+    set => throw new InvalidOperationException();
+  }
 
-  public override IOrigin WithVal(string newVal) {
+  public Token Prev {
+    get => Center.Prev;
+    set => throw new InvalidOperationException();
+  }
+
+  public IOrigin WithVal(string newVal) {
     throw new NotImplementedException();
   }
 
-  public override bool IsSourceToken => !ReferenceEquals(this, NoToken);
+  public bool IsCopy => false;
+
+  public bool IsSourceToken => !ReferenceEquals(this, NoToken);
+  public int kind {
```

Comment:
In a follow-up I will make `IOrigin` not inherit from `Boogie.IToken`, and then these messy overrides can be removed.